rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,869 @@
+
+The `Policy` CRD and the `PolicyRevision` CRD will be updated to include the status conditions.
+See the [Policy and PolicyRevision conditions](#policy-and-policyrevision-conditions) section for more details on how the status conditions will be propagated.
+
+The following `PolicyConditions` will be observed:
+
+- `Scheduled`: The policy was created and is pending to be loaded.
+- `Initialized`: The policy was successfully pulled, precompiled and validated by the leader.
+- `Ready`: The policy was successfully loaded and is ready to be used. This status will include the replica number.
+
+By using the `reason` field in the status condition, we can provide detailed information in case of an error.
+For instance if the leader fails to initialize a policy because of registry error, the following status condition will be set:
+
+```yaml
+conditions:
+  - type: Initialized
+    status: False
+    reason: PullError
+    message: "Cannot pull the policy from the registry."
+    generation: 1
+```
+
+#### Policy phase
+
+Similarly to the `Pod` phase, we will introduce a `phase` field in the `Policy` CRD .
+The phase of a Policy is a simple, high-level summary of where the policy is in its lifecycle.
+
+Here are the possible values for `phase`:
+
+- `Pending`: The policy was created and is pending to be loaded.
+- `Updating`: The policy was updated and is pending to be loaded.
+- `Active`: The policy was successfully loaded and is ready to be used.
+- `Failed`: The policy failed to load.
+
+### Kubernetes Events
+
+The controller will generate [Kubernetes Events](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/) to reflect the status of a policy.
+Events will be created during policy status transitions, such as when a new policy is created, updated, successfully loaded, or encounters an error.
+This enables users to track the policy's status using `kubectl get events`.
+
+### Rollback
+
+If a policy fails to load, the Policy Server will keep running with the previous valid policy.
+The user will be able to rollback to a previous policy version by updating the `Policy` CRD (via a spec field or an annotation, TBD).
+The controller will detect the change and use the `PolicyRevision` CRD to update the Policy CRD with the previous configuration.
+
+### Garbage Collection
+
+When a `Policy` is deleted, the controller will also delete all associated `PolicyRevisions`.
+Using the `OwnerReference` field is not feasible in this case, as the `PolicyRevision` may reside in a different namespace than the `AdmissionPolicy` or `AdmissionPolicyGroup`,
+or it could be versioning a `ClusterAdmissionPolicy` or `ClusterAdmissionPolicyGroup`, which are cluster-scoped resources.
+
+We will also need to implement a garbage collection mechanism to remove old precompiled modules from the shared cache.
+This will prevent the cache from growing indefinitely and consuming unnecessary resources.
```

Comment:
Considering that we will have a configuration to define the number of revision that will store. Don't the garbage collector need to clean conditions from old generations? I guess there is a limit of the number of conditions allowed, isn't it? 

Even if there is not limits for the conditions count, I believe this is a good  thing to have. It's cleaner and easier to the user to process the data with the want to.